File: desktop_app/frontend/src/lib/sandboxEnvironment.ts
export interface SandboxPreloadedItem {
  name: string;
  kind: string;
  description: string;
}

export interface SandboxEnvironmentSpec {
  preloaded: SandboxPreloadedItem[];
  allowed_imports: string[];
  builtins: string[];
  rules: string[];
  toolbar_names: string[];
  starter_snippet: string;
}

export const FALLBACK_SANDBOX_ENVIRONMENT: SandboxEnvironmentSpec = {
  preloaded: [
    { name: 'df', kind: 'dataframe', description: 'Copy of the open dataset' },
    { name: 'result', kind: 'variable', description: 'Assign your final answer here' },
    { name: 'pd', kind: 'module', description: 'pandas' },
    { name: 'np', kind: 'module', description: 'numpy' },
    { name: 'pandas', kind: 'module', description: 'Same as pd' },
    { name: 'numpy', kind: 'module', description: 'Same as np' },
    { name: 'math', kind: 'module', description: 'Standard math functions' },
    { name: 'datetime', kind: 'module', description: 'Date and time utilities' },
  ],
  allowed_imports: ['scipy', 'scipy.stats'],
  builtins: [
    'abs', 'all', 'any', 'bool', 'dict', 'divmod', 'enumerate', 'filter', 'float', 'int',
    'len', 'list', 'map', 'max', 'min', 'pow', 'print', 'range', 'reversed', 'round',
    'set', 'slice', 'sorted', 'str', 'sum', 'type', 'zip',
  ],
  rules: [
    'Assign the final answer to result.',
    'Plotting methods (plot, hist, show, etc.) are not allowed.',
    'File I/O (to_csv, to_json, etc.) is not allowed.',
    'Only modules listed here may be imported.',
  ],
  toolbar_names: ['pd', 'np', 'math', 'datetime', 'df', 'result'],
  starter_snippet:
    'filtered = df[df["column"] > 0]\nresult = filtered.groupby("category").size().reset_index(name="count")',
};

export async function fetchSandboxEnvironment(
  fetchFn: (input: RequestInfo | URL, init?: RequestInit) => Promise<Response>,
  baseUrl: string,
): Promise<SandboxEnvironmentSpec> {
  const response = await fetchFn(`${baseUrl}/sandbox/environment`);
  if (!response.ok) {
    return FALLBACK_SANDBOX_ENVIRONMENT;
  }
  const body = (await response.json()) as { environment?: SandboxEnvironmentSpec };
  return body.environment ?? FALLBACK_SANDBOX_ENVIRONMENT;
}

File: packages/query_execution/query_execution/sandbox_environment.py
"""Structured sandbox environment metadata for API and UI consumers."""

from __future__ import annotations

from typing import Any, Dict, List

from .pandas_code_policy import IMPORT_ALLOW_LIST, PRELOADED_IMPORTS
from .safe_pandas_executor import ALLOWED_BUILTIN_NAMES

_PRELOADED: List[Dict[str, str]] = [
    {"name": "df", "kind": "dataframe", "description": "Copy of the open dataset"},
    {"name": "result", "kind": "variable", "description": "Assign your final answer here"},
    {"name": "pd", "kind": "module", "description": "pandas"},
    {"name": "np", "kind": "module", "description": "numpy"},
    {"name": "pandas", "kind": "module", "description": "Same as pd"},
    {"name": "numpy", "kind": "module", "description": "Same as np"},
    {"name": "math", "kind": "module", "description": "Standard math functions"},
    {"name": "datetime", "kind": "module", "description": "Date and time utilities"},
]

_RULES: List[str] = [
    "Assign the final answer to result.",
    "Plotting methods (plot, hist, show, etc.) are not allowed.",
    "File I/O (to_csv, to_json, etc.) is not allowed.",
    "Only modules listed here may be imported.",
]

_STARTER_SNIPPET = (
    "filtered = df[df[\"column\"] > 0]\n"
    "result = filtered.groupby(\"category\").size().reset_index(name=\"count\")"
)


def sandbox_environment_spec() -> Dict[str, Any]:
    """Return the notebook sandbox environment for desktop UI and docs."""
    preloaded_names = {item["name"] for item in _PRELOADED}
    allowed_imports = sorted(
        module
        for module in IMPORT_ALLOW_LIST
        if module not in PRELOADED_IMPORTS and module not in preloaded_names
    )
    toolbar_names = ["pd", "np", "math", "datetime", "df", "result"]
    return {
        "preloaded": list(_PRELOADED),
        "allowed_imports": allowed_imports,
        "builtins": sorted(ALLOWED_BUILTIN_NAMES),
        "rules": list(_RULES),
        "toolbar_names": toolbar_names,
        "starter_snippet": _STARTER_SNIPPET,
    }
